src/libsyntax_ext/deriving/generic/mod.rs

```diff
@@ -428,8 +428,8 @@ impl<'a> TraitDef<'a> {
                         }
                     }
                     _ => {
-                        cx.span_err(mitem.span,
-                                    "`derive` may only be applied to structs, enums and unions");
+                        // This not-an-ADT case is an error, but it should have
```

Comment:
so this branch should be `bug!("non-ADT derive should have been caught during expansion")`

Reply:
Unfortunately, this will ICE: setting the `span_err` during expansion doesn't stop compilation from reaching this point. (But maybe it should??)

Reply:
Can you add a `debug!` statement here then? Also add a comment pointing at `libsyntax/ext/expand.rs:MacroExpand::expand()` for the handling of this case.

Reply:
> Can you add a debug! statement here then?

_Surprisingly_, no: the `libsyntax_ext` crate doesn't have `debug!` in scope, and when I tried adding `#[macro use] extern crate log;` to lib.rs, we get "can't find crate for `log`". I don't understand what's going on there. (Added the expand filename comment pointer, though.)